torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -89,12 +355,67 @@ class ChunkDataSet final
   }
 
  private:
+  /// running on worker thread to preload chunk data.
+  void preloader(size_t id) {
+    while (!quit_worker_) {
+      try {
+        size_t chunk_id = 0;
+        if (auto chunk_sampler_result = chunk_sampler_.next(1)) {
+          chunk_id = chunk_sampler_result.value()[0];
+        } else {
+          break;
+        }
+        UnwrappedBatchType data = chunk_reader_.read_chunk(chunk_id);
+        AT_CHECK(options_.ignore_empty_chunk_ || !data.empty(),
+                 "Chunk with index " + std::to_string(chunk_id) + " is empty");
+
+        if (data.empty()) {
+          // skip adding the current chunk data and move to the next.
+          batch_buffer_->skip_chunk();
+        }
+        else {
+          batch_buffer_->add_chunk_data(std::move(data));
+        }
+      } catch (...) {
+        batch_buffer_->add_chunk_data(std::current_exception());
+      }
+    }
+  }
+
+  /// Block the current thread until the workers finish execution and exit.
+  void free_workers() {
+    if (!quit_worker_) {
+      quit_worker_ = true;
```

Comment:
here we need to add logic to stop.
      if (chunk_buffer_ != nullptr) {
        chunk_buffer_->stop();
      }